evalml/automl/engine/engine_base.py

```diff
@@ -102,7 +102,23 @@ def submit_scoring_job(
         """Submit job for pipeline scoring."""
 
 
-def train_pipeline(pipeline, X, y, automl_config, schema=True):
+# global hash dictionary to allow minimal calculation of hashes on data splits
```

Comment:
Just some food for thought but is there a way to make this not global via refactoring some methods? Seems like we want it mostly for `train_pipeline` and `train_and_score_pipeline` which are primarily called via the Engine classes... could we somehow encapsulate these functions in the Engine classes and then have the hash_dict be a class attribute?

Reply:
another option is having `AutoMLSearch` hold the cache and pass it as part of `automl_config` or another parameter.

Reply:
I did try that initially! I basically hit a point where I wasn't sure the best way to approach this without completely refactoring the methods and changing the args that we pass to and from the engines to the `train_and_score_pipelines` method. Using global made the code a lot more manageable, which was why I went with this approach. If anyone has suggestions on how I can refactor instead, would love to hear it!

Reply:
@bchen1116 The `hash_dict` is only used to keep track of whether or not we've hashed the data previously right? Is that still a concern now that we're only hashing the index as opposed to the entire data?

I did some back-of-the-envelope calculations and hashing an index of 10 million entries, 3 times for every pipeline, and for 50 pipelines would take 15 seconds. Since this assumes a 3-fold cross-val, this means the total search time increase for 30 million rows is 15 seconds. I think that is a small fraction of time compared to the total search time. The speed up for KDDCup (less than 30 million rows) is much bigger than 15 seconds.

I vote we get rid of this optimization if possible. If we want to keep it for max performance, I think we should brainstorm ways to refactor this. Can we store the `hash_dict` in the engine instance as opposed to AutoMLSearch?